nemo/collections/speechlm/data/data_sampler.py

```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from torch.utils.data import DataLoader, IterableDataset
+from nemo.lightning.pytorch.plugins import MegatronDataSampler
+from nemo.utils import logging
+
+
+class SpeechLMDataSampler(MegatronDataSampler):
+    """
+    Overwrite the default MegatronDataSampler to not add batch sampler when iteralbe dataset is used.
+    """
+
+    def transform_dataloader(self, dataloader: DataLoader, consumed_samples: int = 0) -> DataLoader:
+        """
+        Overwrites the MegatronDataSampler.transform_dataloader() function
+        """
+        sampler = getattr(dataloader, 'sampler', None)
+
+        if sampler is not None and "lhotse." in str(type(sampler)):
+            logging.info(f"Using Lhotse sampler for dataloader {dataloader}. Skipping Megatron data sampler.")
```

Comment:
`MegatronDataSampler` seems to be implementing the logic for splitting batches into micro batches based on some predefined settings. If we completely skip it, and don't add our own logic for the splitting, does that mean lhotse based dataloading doesn't work with any model parallelism at all? Was that tested?

Again if we maintain two dataloader implementations, we will have to test both of them against the full feature set we want to support. AFAIK at the very least TP should be supported for SpeechLM in NeMo 2.

Reply:
TP support is verified with both lhotse as well, we did the same thing in NeMo 1.x by not using Megatron sampler, and it worked.